### PR TITLE
Sparse rework: nalgebra-sparse

### PR DESCRIPTION
**Pull request by Andlon**
This PR represents a major rework, redesign and extension of nalgebra's sparse capabilities. It effectively addresses #592, #234, and eventually makes #606 redundant.

I've tried to present a relatively polished product here, but there are definitely open questions and problems, some of which I'd like to leave for future PRs. I've tried to summarize the most significant points below.

I invite anyone who's interested in sparse matrix support in `nalgebra` to take a look and let us know if the functionality would help address your needs, or if you have concerns about the design or implementation!

# Summary

- Some nalgebra contributions: A new module `proptest` under the `proptest-support` feature. Here we provide some methods and types to automatically generate matrices and vectors that integrate with `proptest` and provide automated invariants-preserving shrinking. This constitutes the first 8 commits: f9ea2b4471101fff9e53844a964a5ba5a58dbc75 through 646f62a293a878b66651c4ce4855c16237cf306b.
- New crate `nalgebra-sparse` added to the `nalgebra` workspace. The remaining commits are concerned with `nalgebra-sparse`.
	- Implements CSR, CSC and COO matrices, along with conversions and arithmetic operations.
	- Provides `SparsityPattern`, an abstraction for sparsity patterns of CSR/COO matrices. After long deliberation, I eventually ditched the `Arc<SparsityPattern>` design proposed in #592. Although the `Arc` design provides some notable benefits, I eventually came to the conclusion that the memory savings and potential fast paths are not worth the slightly added complexity in the future when integrating with existing matrix libraries (e.g. Intel MKL). The sparsity patterns of CSR and CSC matrices are still stored and encoded in instances of `SparsityPattern`, however.
	- Fairly well tested with extensive use of proptests.
- Some basic CI instructions for testing `nalgebra-sparse` added to the Circle CI config.

# How to review this PR

This PR is arguably rather enormous. I *strongly* recommend to start by browsing the docs, so you get an overview of the design and philosophy:

```
# Docs for the new nalgebra `proptest` contributions. See the `proptest` module in nalgebra
cargo doc --features proptest-support --open

# Docs for the new sparse contributions
cargo doc --manifest-path=nalgebra-sparse/Cargo.toml --features proptest-support,compare --open
```

I encourage you to note down any issues that you see. Small fixes I will apply directly, larger ones we can discuss whether to postpone to a future PR or if it's urgent enough to implement in this initial PR still.

Also, please check out the CI config for how to run tests. There are some new additions here, such as the addition of a `slow-tests` feature flag (for both `nalgebra` and `nalgebra-sparse`) to run some particularly long-running tests.

# Open questions, known issues etc.

- I did not add a separate README, nor have I modified nalgebra's README or nalgebra's documentation. How should we "advertise" the new sparse functionality?
- Some issues with `proptest`:
	- Proptest is relying on an older version of rand, causes some complaints about incompatibility
	  by e.g. cargo. There's a PR (AltSysrq/proptest/#220) that addresses this, but it has yet to be merged.
	- There's an issue with the `Strategy::prop_shuffle` method. There's a PR that fixes this (AltSysrq/proptest/#217) but it has yet to be merged. I hacked my way around this by including a fixed but stripped-down version of the `Shuffle` strategy in `proptest`. It is isolated from the rest of the code and clearly commented as a stopgap solution. As soon as this is fixed in `proptest` we can drop this code.
- Trait bounds for arithmetic operations are all over the place. It would be nice to somehow agree to something consistent here. The thing is that I don't want to, say, only require `Sub` for `Sub<CsrMatrix> for CsrMatrix`, because I want to be able to call computational kernels from `crate::ops::serial` in the implementations of `Add, Sub` etc. @sebcrozet and I have briefly discussed the possibility of including something like a `Number` trait in `nalgebra` on a separate occasion, which would have all the usual arithmetic operations except the multiplicative identity. Perhaps this could be relevant here...
- There's quite a bit of repetition/duplication between CSR and CSC, especially in tests. This could perhaps be reduced, but at the cost of additional complexity, and I'm not convinced that this is worth it.
- Ideally we should have some more simple unit tests to complement many of the property-based tests, just as sanity tests. However, this would require a lot of time to do, it is extremely boring and in the end it would with almost 100% certainty just confirm that things are already working. So I'm not entirely sure if this is ultimately worthwhile.
- Consistency: I've tried to use method names mostly consistent with `nalgebra`, but I might have missed some opportunities here, so I'd appreciate if you call out names that are not consistent with `nalgebra`. One thing that I recently noticed: I've been using methods like `.col(i)` whereas `nalgebra` uses `.column(i)`. I have to admit I've never been a big fan of `nalgebra`'s naming in this regard, because it's somewhat unsymmetric with respect to rows: `.row(i)` is much shorter than `.column(i)`, so `.col(i)` would not be too bad.... In any case, perhaps I should update `nalgebra-sparse` to consistently use `column` everywhere?
- Should we export `nalgebra` from the `nalgebra-sparse` crate root?

## Missing functionality (future PRs?)

I think I want to address the following in future PRs:

- Serde integration (cannot just derive blindly as we need to validate the input to ensure that the implementation is sound even if the serialized data is faulty).
- Matrixmarket IO. This is already implemented for `nalgebra/sparse`, but I'd like to maybe keep it out of this PR in order to not further bloat the scope.
- The conversion functions and "low-level" arithmetic ops will probably receive a "workspace" argument in the future, to give more control over intermediate allocations.
- Support borrowed CSR/CSC data and other index types than `usize` (typically 32-bit is sufficient, which might give drastic performance improvements over `usize` on 64-bit improvements).
- What's missing right now is a method to construct `CsrMatrix/CscMatrix` from unsorted but otherwise valid CSR/CSC data (i.e. the column indices might not be sorted within each row). This seems crucial for interopability with some matrix libraries that don't necessarily enforce this property.
- Performance! This includes benchmarking, optimizing existing implementations and ultimately also adding parallel routines.

## Maintenance

I am not planning to take my leave after we get this PR merged. It makes sense for me to be involved in - and perhaps even responsible for - the maintenance and further development of `nalgebra-sparse`. It would be good to discuss in what form this should take place. For example, I could take on the main responsibility for reviewing PRs and organizing issues related to `nalgebra-sparse`.

## Acknowledgements

I'd like to especially thank @w1th0utnam3 for fruitful discussions and feedback, and @sebcrozet for several discussions on Discord.

### CODE REVIEW

**Comment by sebcrozet** [comment]
Great work :100: Thank you for this PR. I will start reviewing next week.

**Comment by Andlon** [comment]
Ping @elrnv, let me know if you have time to take a look and have any concerns about the current design and trajectory :-)

**Review by elrnv — COMMENTED** [review]
First of all, thank you! What a huge addition to an already extensive library!
I am also grateful for the extensive documentation of this PR. Making the review was a pleasure!

I am already excited to try out some sparse algorithms with this :)

The PR already looks like it is in great shape. I only have a few minor suggestions.

In response to the Open questions section:

- > Ideally we should have some more simple unit tests to complement many of the property-based tests,
  > just as sanity tests. However, this would require a lot of time to do, it is extremely boring and
  > in the end it would with almost 100% certainty just confirm that things are already working. So
  > I'm not entirely sure if this is ultimately worthwhile.
  
   * I wholeheartedly agree that this is boring, but I would like to stress that small unit tests also
  help other contributors catch bugs in changes to your code later on. I think this will be especially true when the sparse system gets larger and more complex.

- > Consistency...
  * I want to second this. In my experience `col` is fairly unambiguous and ubiquitous in this context.  Perhaps having `col` as an alias to `column` would work throughout `nalgebra`?

API review:

- The main example in the docs shows how to populate a COO matrix by using `new` and `push`. If this  is the preferred method to populate a matrix, shall we also add a `reserve` function to allocate additional space (for all of values, row indices and column indices).

- I would also like to suggest an alternative to `nnz`. Although this is a common acronym, it is not  all that descriptive in itself and can possibly hurt readability in code. Perhaps `non_zeros` or `num_non_zeros` would be more descriptive? Also I expect `triplet_iter` would be more commonly  used, so having a short name here may not be as important. On the other hand this will be less  consistent with `ncols` and `nrows`. Anyways it seems like something that should be decided now to  avoid future breaking changes.

- It was mentioned that SparsityPattern was stored in an `Arc`, which would be great IMO. Do you think it's possible to make SparsityPattern generic in the same way that Matrix storage is in nalgebra? This can perhaps allow users to chose `Arc<SparsityPattern>` if they like or even `Rc<SparsityPattern>`. Perhaps this is best left to another PR if at all.

- A very minor ask: could we move the matrix constructors towards the beginning of the impl so they  are more easily discoverable in the docs? I am thinking particularly of`identity` but perhaps a  `diagonal` constructor could be added. In general I found it helpful in other libs to see all  methods that output `Self` in one place.

Implementation high level review:

- In the spirit of the comprehensive docs of nalgebra and nalgebra-sparse, it would be a welcome  contribution to see an example for Cholesky factorization and symbolic factorization to outline   the benefit of using `CscSymbolicCholesky` to precompute the symbolic sparsity pattern instead of  just using `refactor` on `CscCholesky`. I think this would be helpful under
  `factor_numerical` and `CscSymbolicCholesky` (probably the same example would work in both places).
  ```
  let col_offsets = vec![0, 2, 4, 5, 6];
  let row_indices = vec![0, 3, 0, 2, 2, 1];
  let values = vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0];
  let csc = CscMatrix::try_from_csc_data(4, 4, col_offsets, row_indices, values)
    .expect("CSC data must conform to format specifications");
  let sparsity = csc.pattern().clone();
  let symbolic_cholesky = CscSymbolicCholesky::factor(sparsity);

  let cholesky = CscCholesky::factor(
    symbolic_cholesky.clone(),
    csc
  );

  // ... do a solve perhaps
  ```
  On an additional note, there is a lot of cloning of the sparsity pattern going on in this example.     Perhaps it is a good candidate to test if there are any performance benefits in having the sparsity pattern behind an `Arc`.

File: nalgebra-sparse/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+//! nalgebra_sparse = "0.1"
+//! ```
+//!
+//! # Supported matrix formats
+//!
+//! | Format                  | Notes                                        |
+//! | ------------------------|--------------------------------------------- |
+//! | [COO](`coo::CooMatrix`) | Well-suited for matrix construction. <br /> Ill-suited for algebraic operations. |
+//! | [CSR](`csr::CsrMatrix`) | Immutable sparsity pattern, suitable for algebraic operations. <br /> Fast row access. |
+//! | [CSC](`csr::CscMatrix`) | Immutable sparsity pattern, suitable for algebraic operations. <br /> Fast column access. |
```
[the remark this answers is not in the document]
Comment (Andlon) [reply nalgebra-sparse/src/lib.rs:58]: Ah, well spotted, thanks!

File: nalgebra-sparse/src/csc.rs [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+
+    /// Returns a reference to the sparsity pattern and a mutable reference to the values.
+    #[inline]
+    pub fn pattern_and_values_mut(&mut self) -> (&SparsityPattern, &mut [T]) {
+        self.cs.pattern_and_values_mut()
+    }
+
+    /// Returns the underlying sparsity pattern.
+    ///
+    /// The sparsity pattern is stored internally inside an `Arc`. This allows users to re-use
```
[the remark this answers is not in the document]
Comment (Andlon) [reply nalgebra-sparse/src/csc.rs:374]: Indeed, good catch!

**Comment by Andlon** [comment]
> First of all, thank you! What a huge addition to an already extensive library!
> I am also grateful for the extensive documentation of this PR. Making the review was a pleasure!
> 
> I am already excited to try out some sparse algorithms with this :)

Happy to hear that :)

> I wholeheartedly agree that this is boring, but I would like to stress that small unit tests also
> help other contributors catch bugs in changes to your code later on. I think this will be especially true when the sparse system gets larger and more complex.

I entirely agree! I appreciate that you pointed this out, because I realize what I wrote there wasn't so clear. So almost everything is either already explicitly or implicitly tested - though I've probably missed a few things. Actually when I wrote this it strikes me that I probably never tested the `identity` methods for `CsrMatrix`/`CscMatrix`, so I should look into this and make another commit... Anyway, what I meant with my sloppy comment is that I have tests like these:

```rust
#[test]
fn csr_transpose_agrees_with_dense(csr in csr_strategy()) {
    let dense_transpose = DMatrix::from(&csr).transpose();
    let csr_transpose = csr.transpose();
    prop_assert_eq!(dense_transpose, DMatrix::from(&csr_transpose));
    prop_assert_eq!(csr.nnz(), csr_transpose.nnz());
}
```
If you're not familiar with `proptest`, this test is run many times for randomized input defined by `csr_strategy()` (by default 10000 on CI, I sometimes run them for hundreds of millions of samples on my local machine). In this case it generates CSR matrices with reasonably sized integer values and rows and columns both in the range `0..=6`. This particular verifies that for every `csr` matrix in our sample space (statistically speaking), these two properties hold:

1. That `DMatrix::from(&csr.transpose()) == DMatrix::from(&csr).transpose()`, i.e. the *dense representation* of the CSR matrix transposed is equal to the transpose of the dense representation.
2. That the `csr.transpose()` has the same number of non-zeros as `csr`.

The second property is necessary because `csr.transpose()` could satisfy property (1) while still adding additional unwanted fill-in to the matrix (i.e. explicit zeros). In any case, these two properties together essentially *define* what a correct implementation is. So if this tests passes, we're almost guaranteed to have a correct implementation (barring any very particular corner cases that are not caught by the sample space or are very unlikely to be generated). Clearly, this is simultaneously:

1. superior to unit tests, because we don't test a small number of input-output pairs, but a very large number of inputs (and outputs by virtue of whether they satisfy the defining properties).
2. much more fun to write than a unit test, because we don't have to come up with input data and compute the correct result
3. far more useful for development, as in addition to getting randomized input data, `proptest` will also "shrink" the input for us, so more often than not you will get very "simple" examples that demonstrate test failures (for example, a failing 6x6 matrix might be shrunk to a 2x2 matrix, so that it's much easier to debug).

So, what's the catch? Well, what if `csr_strategy()` is broken? What if it just always returns the same matrix? Then our whole argument falls apart, and the test might give us a false sense of security. Now, I've taken steps to ensure that this is not the case. Looking at the implementation of `csr_strategy()`, we have:

```rust
fn csr_strategy() -> impl Strategy<Value = CsrMatrix<i32>> {
    csr(-5..=5, 0..=6usize, 0..=6usize, 40)
}
```

so basically it just calls `nalgebra_sparse::proptest::csr`. There's a test (not a `proptest`) called `csr_samples_all_admissible_outputs` that tests that for a small row and column range that `csr(...)` indeed *does* generate all possible values in the admissible space of outputs. So with this test we can feel reasonably sure that our `csr_strategy()` is not broken. *However*, many tests require various transformations of the strategies, for example this one:

```rust
#[test]
fn csr_mul_dense(
    // a and b have dimensions compatible for multiplication
    (a, b)
    in csr_strategy()
        .prop_flat_map(|a| {
            let cols = PROPTEST_MATRIX_DIM;
            let b = matrix(PROPTEST_I32_VALUE_STRATEGY, a.ncols(), cols);
            (Just(a), b)
        }))
{ ... }
```
And since it would be very laborious to test every little such transformation, it's possible that various mistakes sneak in, or that some part of the machinery is subtly broken, so that we don't sample a large enough input space. 
I usually manually inspect the many of the inputs/outputs generated when writing these tests, but of course, that's no guarantee that things won't change in the future.

That's why I think it would be nice to have some additional unit tests with basic input, just as an additional "sanity test", to make sure that *if* something goes wrong with any of the `proptest` machinery, then at least we have some basic unit tests that ensure that our code is at least not *completely* wrong.

So, going back to your original point, the current tests indeed do prevent regressions. Or, at least, they should, if we're willing to test the randomized inputs generated ;-)

> I want to second this. In my experience `col` is fairly unambiguous and ubiquitous in this context.  Perhaps having `col` as an alias to `column` would work throughout `nalgebra`?

I have thought about that too, and personally I think that would be nice. Let's see what @sebcrozet thinks about it.

> The main example in the docs shows how to populate a COO matrix by using `new` and `push`. If this  is the preferred method to populate a matrix, shall we also add a `reserve` function to allocate additional space (for all of values, row indices and column indices).

Absolutely! I think I'll prefer to add this in a subsequent PR, however, as it's a performance optimization.

>  I would also like to suggest an alternative to `nnz`. Although this is a common acronym, it is not  all that descriptive in itself and can possibly hurt readability in code. Perhaps `non_zeros` or `num_non_zeros` would be more descriptive?

Interesting perspective. I guess for newcomers `nnz` might not be so descriptive, indeed. On the other hand, it's ubiquitious in literature on sparse matrices. I think `non_zeros` might be misunderstood to mean *the non-zero values themselves*, i.e. what `values()` does at the moment, so I think then we should rather use something like `num_non_zeros`. On the other hand, that's rather long for something that comes up very frequently (because often you want e.g. auxiliary data structures to have the same size). This is one of the things that might be reasonable to "teach" to newcomers. Similar to how they need to learn what CSR or COO is, nnz is one more acronym. Though we should keep the number of acronyms to a minimum... I am not decided here, perhaps other also have some thoughts on this.

>  Also I expect `triplet_iter` would be more commonly  used, so having a short name here may not be as important. On the other hand this will be less  consistent with `ncols` and `nrows`. Anyways it seems like something that should be decided now to  avoid future breaking changes.

Sorry, I don't understand, could you maybe clarify?

> It was mentioned that SparsityPattern was stored in an `Arc`, which would be great IMO. Do you think it's possible to make SparsityPattern generic in the same way that Matrix storage is in nalgebra? This can perhaps allow users to chose `Arc<SparsityPattern>` if they like or even `Rc<SparsityPattern>`. Perhaps this is best left to another PR if at all.

So the main reason I left it out was because it is not so clear what the right way of making it work with *borrowed data* is. There are some nuances there that I simply decided to postpone. Indeed, I do intend to replace `CsrMatrix<T>` with something like `struct CsrMatrix<T, Storage=DefaultCsrStorage<T>>`, which could eventually let us plug `Arc` back in as an "alternative" storage (or even the default). However, it's very important to me that the syntax and docs don't become overly generic/complicated. I suggest we revisit this in a future issue/PR! We should see how it interacts with generalizing the storage to allow borrowed data.


> A very minor ask: could we move the matrix constructors towards the beginning of the impl so they  are more easily discoverable in the docs? I am thinking particularly of`identity` but perhaps a  `diagonal` constructor could be added. In general I found it helpful in other libs to see all  methods that output `Self` in one place.

Ah, I didn't even think about this at all. I definitely agree though! I would prefer to do this in a follow-up PR, I think, because it sounds like it could take a little bit of time, and I need to attend to some other under-attended projects for a while ;-)

> In the spirit of the comprehensive docs of nalgebra and nalgebra-sparse, it would be a welcome  contribution to see an example for Cholesky factorization and symbolic factorization to outline   the benefit of using `CscSymbolicCholesky` to precompute the symbolic sparsity pattern instead of  just using `refactor` on `CscCholesky`. I think this would be helpful under
>       `factor_numerical` and `CscSymbolicCholesky` (probably the same example would work in both places).
> ```
> let col_offsets = vec![0, 2, 4, 5, 6];
> let row_indices = vec![0, 3, 0, 2, 2, 1];
> let values = vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0];
> let csc = CscMatrix::try_from_csc_data(4, 4, col_offsets, row_indices, values)
>                  .expect("CSC data must conform to format specifications");
> let sparsity = csc.pattern().clone();
> let symbolic_cholesky = CscSymbolicCholesky::factor(sparsity);
>      
> let cholesky = CscCholesky::factor(
>     symbolic_cholesky.clone(),
>     csc
> );
>       
> // ... do a solve perhaps
> ```
>       
>       
> On an additional note, there is a lot of cloning of the sparsity pattern going on in this example.     Perhaps it is a good candidate to test if there are any performance benefits in having the sparsity pattern behind an `Arc`.

I have to admit I deliberately did not spend too much time on the sparse Cholesky factorization. It's a port from `nalgebra`'s `sparse` module, and I included it mostly to have some level of almost-feature-parity with `nalgebra/sparse`, so that we can eventually deprecate it (we're still missing matrixmarket IO though). As I understand it, @sebcrozet experimented with it for use with soft bodies in `nphysics`, but ultimately ended up not using it there, so although it works, it's missing some crucial functionality, such as fill-in reduction, to make it truly useful. With more time, I'd love to implement fill-in reduction and make it more generally usable...

Moreover, I think at the moment there's not *that* much benefit to separate the symbolic and numerical factorization, since `CscCholesky::refactor` allows you to only perform the numerical re-factorization while reusing the symbolic factorization, and I imagine most users will not need to create a separate symbolic factorization that only depends on the sparsity pattern. Perhaps the documentation is not so clear on this, however, and could be improved?

----

Thank you very much again for your feedback! I will directly address the smaller things. For the things that we might postpone until later, I will make separate issues, so that they can be addressed at a later time. I'd also be interested to hear your thoughts on the unit test question (and any other things I have not fully addressed!).

**Comment by elrnv** [comment]
@Andlon thanks for the response! I did notice that the proptest stuff that is already included is comprehensive, but I am not that familiar with proptest. I didn't mean to say that there needs to be more testing. I just wanted to emphasize that unit testing is useful in more ways than just making sure the code works now. I am not a TDD diehard, but I did find that sprinkling unit tests throughout development (like for code paths that can get tricky e.g. filtering or iteration) help find problems very quickly. For well designed unit tests I can often find bugs just by looking at the name of the failed unit tests.

So I think given that the current testing is already pretty extensive, it makes sense to just leave adding specific unit tests for future contributors if/where they would find them helpful :)

**Comment by Andlon** [comment]
> @Andlon thanks for the response! I did notice that the proptest stuff that is already included is comprehensive, but I am not that familiar with proptest. I didn't mean to say that there needs to be more testing. I just wanted to emphasize that unit testing is useful in more ways than just making sure the code works now. I am not a TDD diehard, but I did find that sprinkling unit tests throughout development (like for code paths that can get tricky e.g. filtering or iteration) help find problems very quickly. For well designed unit tests I can often find bugs just by looking at the name of the failed unit tests.

I think we're pretty much in agreement here!

> So I think given that the current testing is already pretty extensive, it makes sense to just leave adding specific unit tests for future contributors if/where they would find them helpful :)

Yeah, I hope that's sufficient. I would still *ideally* prefer to write some additional basic input-output unit tests (although I'm almost entirely certain that anything these tests can uncover are already covered by the property tests), but it's too time consuming for me to justify the time spent on this at the moment. Note that many parts of e.g. the `CsrMatrix/CscMatrix` APIs are already tested with ordinary unit tests in addition to property-based tests, however.

I hope it should be sufficient for contributors of new functionality to only test the new functionality, but of course we'll have to see on a case-by-case basis. For example, most of my tests only test `T = i32` because the implementation is generic over `T`. However, if we were to e.g. specialize for floating-point types, then we'd need additional tests.

**Comment by Andlon** [comment]
Currently, `CsrMatrix`, `CscMatrix` and `CooMatrix` are each located in their own top-level modules `csr`, `csc` and `coo`. I've been thinking that maybe it might be better to organize it as follows:

```
formats::{CsrMatrix, CscMatrix, CooMatrix}
formats::csr // Holds all auxiliary CSR structures, such as iterators etc.
formats::csc // Holds all auxiliary CSC structures
formats::coo // Holds all auxiliary COO structures
```

Thus the idea would be to have the sparse matrix formats themselves be accessible in `formats`, and auxiliary data structures in format-specific submodules. This has the benefit of avoiding cluttering of the top-level module if we add more formats (in particular, I plan to add Bcsr (block CSR) and possibly some kind of diagonal matrix format in the future).

**Review by sebcrozet — COMMENTED** [review]
Thank you for this huge PR! And thank to @elrnv for making a first review.

This is a very high-quality PR with great documentation API design. I don't have much to say that needs to be absolutely changed in this first PR.

At first I thought that in order to make the API of `nalgebra-sparse` coherent with `nalgebra`, the function from `ops::serial` should be changed to be methods of `CscMatrix, etc.`  However, I guess this would become a problem for functions like `spmm_csr_dense` because of orphan rules. So to keep the API of `nalgebra-sparse` itself clean, I guess we don't have any choice but keep using functions instead of methods.

> > I want to second this. In my experience col is fairly unambiguous and ubiquitous in this context. Perhaps having col as an alias to column would work throughout nalgebra?
>
> I have thought about that too, and personally I think that would be nice. Let's see what @sebcrozet thinks about it.

I'm OK with adding a `.col` method to nalgebra and deprecating `.column`. I agree that `column` is quite verbose compared to `.row`.

> Interesting perspective. I guess for newcomers nnz might not be so descriptive, indeed. On the other hand, it's ubiquitious in literature on sparse matrices. I think non_zeros might be misunderstood to mean the non-zero values themselves, i.e. what values() does at the moment, so I think then we should rather use something like num_non_zeros.

If `nnz` is the de-facto name given to the number of stored values, then we should keep it. In order to make things easier to newcomers I would suggest adding another methods called `len` that does the same think as `nnz`. In all Rust collections `.len()` returns the number of elements stored in the collection. So I think this makes sense here, and it's the first method a newcomer will use for this.

> Ah, I didn't even think about this at all. I definitely agree though! I would prefer to do this in a follow-up PR, I think, because it sounds like it could take a little bit of time, and I need to attend to some other under-attended projects for a while ;-)

@Andlon No need to bother you with this. I will take care of this myself before merging the PR.

> ... SparsityPattern was stored in an Arc ...

Since this is an optimization I agree this should be on a separate PRs, with some benchmarks supporting the change.

> Thus the idea would be to have the sparse matrix formats themselves be accessible in formats, and auxiliary data structures in format-specific submodules. This has the benefit of avoiding cluttering of the top-level module if we add more formats (in particular, I plan to add Bcsr (block CSR) and possibly some kind of diagonal matrix format in the future).

Right now I think the current `csc`, `csr`, `coo` modules are OK. I would actually be in favor of re-exporting `csc::CscMatrix`, `csr::CsrMatrix`, `csr::CsoMatrix`. to the top-level of the crate. This would be coherent with the may `nalgebra` re-exports all the matrix/vector/transformation types at the top-level of the crate.

Here is a summary of the changes I'm suggesting here:
- Add a `.len()` method with is an alias for `.nnz()`.
- Re-export `csc::CscMatrix`, `csr::CsrMatrix`, and `csr::CsoMatrix` at the root of `nalgebra-sparse`.
- When possible, move the constructors at the top of the methods lists in the docs.

@Andlon If you are OK with these suggestions (or some of them) I can apply them myself before merging the PR if this can save you some time. I will make some changes anyway to make the CI pass for `#[no-std]` environments.

**Comment by sebcrozet** [comment]
About re-exporting `nalgebra` from `nalgebra-sparse`: yes, I think this would be good.
About advertisement:
- `nalgebra-sparse` will be mentioned on the `nalgebra.org` front-page and on its user-guide.
- There will be a post on `r/reddit`, and a tweet once `nalgebra-sparse` is released.
- We should add a section to the README mentioning `nalgebra-glm` and `nalgebra-sparse` (no need to do this right now on this PR though).

**Comment by Andlon** [comment]
@sebcrozet:

thanks for your feedback!

> At first I thought that in order to make the API of `nalgebra-sparse` coherent with `nalgebra`, the function from `ops::serial` should be changed to be methods of `CscMatrix, etc.` However, I guess this would become a problem for functions like `spmm_csr_dense` because of orphan rules. So to keep the API of `nalgebra-sparse` itself clean, I guess we don't have any choice but keep using functions instead of methods.

Yes, it is correct that for consistency, these functions need to remain free functions due to the orphan rules. An additional motivation was to keep the API of `CsrMatrix`, `CscMatrix` and so on free of these "advanced" methods, so that the API surface for e.g. `CsrMatrix` is kept smaller and the docs more accessible to newcomers who are unlikely to need these methods, at least in the beginning.

> > > I want to second this. In my experience col is fairly unambiguous and ubiquitous in this context. Perhaps having col as an alias to column would work throughout nalgebra?
> > 
> > 
> > I have thought about that too, and personally I think that would be nice. Let's see what @sebcrozet thinks about it.
> 
> I'm OK with adding a `.col` method to nalgebra and deprecating `.column`. I agree that `column` is quite verbose compared to `.row`.

Sounds nice!

> > Interesting perspective. I guess for newcomers nnz might not be so descriptive, indeed. On the other hand, it's ubiquitious in literature on sparse matrices. I think non_zeros might be misunderstood to mean the non-zero values themselves, i.e. what values() does at the moment, so I think then we should rather use something like num_non_zeros.
> 
> If `nnz` is the de-facto name given to the number of stored values, then we should keep it. In order to make things easier to newcomers I would suggest adding another methods called `len` that does the same think as `nnz`. In all Rust collections `.len()` returns the number of elements stored in the collection. So I think this makes sense here, and it's the first method a newcomer will use for this.

Heh, to be honest I'm not entirely sure how I feel about that. I guess it depends on your perspective. You can think of a sparse matrix data structure in at least two different ways:

- a data structure storing `nnz` entries. In this case `.len()` would make sense.
- a **matrix** *that just happens to be explicitly storing `nnz` entries internally*. Calling `.len()` on a matrix feels... icky.

I suppose I'm in the second camp, and I feel that `.len()` would be semantically... strange? I welcome the suggestion as a middle-ground, but I'm a little wary of straying from "linear algebra terminology" in order to seem more similar to Rust collections.

I also didn't realize before now that `nalgebra` defines `.len()` for *all* matrices, not just vectors. The docs for `nalgebra::Matrix::len` say:

> The total number of elements of this matrix.

If we were to define `nnz == len` for sparse matrices, then it seems that this would not be consistent with the `nalgebra` definition of `.len()`.

> > Ah, I didn't even think about this at all. I definitely agree though! I would prefer to do this in a follow-up PR, I think, because it sounds like it could take a little bit of time, and I need to attend to some other under-attended projects for a while ;-)
> 
> @Andlon No need to bother you with this. I will take care of this myself before merging the PR.

That's great, thanks!
 
> Right now I think the current `csc`, `csr`, `coo` modules are OK. I would actually be in favor of re-exporting `csc::CscMatrix`, `csr::CsrMatrix`, `csr::CsoMatrix`. to the top-level of the crate. This would be coherent with the may `nalgebra` re-exports all the matrix/vector/transformation types at the top-level of the crate.

I think that makes sense. I was playing around with this at some point as well, not sure why I decided against it. Perhaps it was different with an earlier API design. Anyway, since users are likely to mostly work with the matrix formats, it makes sense to make them readily accessible like that.

> Here is a summary of the changes I'm suggesting here:
> 
>     * Add a `.len()` method with is an alias for `.nnz()`.
> 
>     * Re-export `csc::CscMatrix`, `csr::CsrMatrix`, and `csr::CsoMatrix` at the root of `nalgebra-sparse`.
> 
>     * When possible, move the constructors at the top of the methods lists in the docs.
> 
> 
> @Andlon If you are OK with these suggestions (or some of them) I can apply them myself before merging the PR if this can save you some time. I will make some changes anyway to make the CI pass for `#[no-std]` environments.

I think the `.len()` question might need a little more thought/discussion, but I'd definitely appreciate it if you would apply these suggestions! That would certainly take a load off my shoulders, since I'm trying to wrap up a number of different things before I go on leave for ~2 months.

**Comment by Andlon** [comment]
> About re-exporting `nalgebra` from `nalgebra-sparse`: yes, I think this would be good.
> 
> About advertising `nalgebra-sparse`:
> 
> - `nalgebra-sparse` will be mentioned on the `nalgebra.org` front-page and on its user-guide.
> - There will be a post on `r/reddit`, and a tweet once `nalgebra-sparse` is released.
> - We should add a section to the README mentioning `nalgebra-glm` and `nalgebra-sparse` (no need to do this right now on this PR though).

Sounds great! Perhaps also consider adding a note on `nalgebra-glm` and `nalgebra-sparse` to the `nalgebra` API docs, so that they are discoverable to users browsing e.g. `docs.rs`. I don't know if other people are like me, but I almost exclusively use `docs.rs` to look at new crates - often I end up not even looking at the README this way.

**Comment by sebcrozet** [comment]
I applied the changed we discussed.
Note that I had to remove the `proptest` dev-dependency because it triggered a [known Cargo bug](https://github.com/rust-lang/cargo/issues/4866) that breaks `#[no-std]` builds.

**Comment by sebcrozet** [comment]
Thank you again for all this work @Andlon!

**Comment by Andlon** [comment]
I'm back from my hiatus, and as I promised I've made issues for most of the remaining points that were not addressed by this PR. Let me know if I've missed something!

#876, #877, #878, #879, #880, #881, #882
